src/lib/comms/sol-lwm2m.c

```diff
@@ -3014,9 +3014,106 @@ clear_bootstrap_ctx(struct sol_lwm2m_client *client)
+    struct sol_lwm2m_resource res[2];
+    int r;
+    bool ret = false;
+    int64_t default_acl = SOL_LWM2M_ACL_NONE;
+    uint16_t i;
+
+    SOL_WRN("connections.len=%" PRIu16, client->connections.len);
+
+    //If only one server or Bootstrap Server ID, then full access rights
+    if (client->connections.len == 1 || server_id == UINT16_MAX)
+        return true;
+
+    SOL_WRN("server_id=%" PRId64 " obj=/%" PRIu16 "/%" PRId32 " needed=%" PRId64,
+        server_id, obj_id, instance_id, rights_needed);
+
+    obj_ctx = find_object_ctx_by_id(client, ACCESS_CONTROL_OBJECT_ID);
+    SOL_VECTOR_FOREACH_IDX (&obj_ctx->instances, obj_instance, i) {
+        r = read_resources(client, obj_ctx, obj_instance, res, 2,
+            ACCESS_CONTROL_OBJECT_OBJECT_RES_ID,
+            ACCESS_CONTROL_OBJECT_INSTANCE_RES_ID);
+        if (r < 0) {
+            SOL_WRN("Could not read Access Control Object's"
+                " [Object ID] and [Instance ID] resources\n");
+            continue;
+        }
+
+        //Retrieve the associated Access Control Object Instance, by matching Object ID
+        // and Instance ID, or if instance_id == -1 and 'R'ead is needed, this
+        // is an Observe Request on Object level and any Instance with this access right
+        // is enough to reply with the Observe falg on CoAP
+        if ((res[0].data[0].content.integer == obj_id &&
+            res[1].data[0].content.integer == instance_id) ||
+            (res[0].data[0].content.integer == obj_id &&
+            instance_id == -1 && (rights_needed & SOL_LWM2M_ACL_READ))) {
+            clear_resource_array(res, sol_util_array_size(res));
+
+            r = read_resources(client, obj_ctx, obj_instance, res, 2,
+                ACCESS_CONTROL_OBJECT_ACL_RES_ID,
+                ACCESS_CONTROL_OBJECT_OWNER_RES_ID);
+            if (r < 0) {
+                SOL_WRN("Could not read Access Control"
+                    " Object's [ACL] and [Owner ID] resources\n");
+                continue;
+            }
+
+            //Retrive this server's ACL Resource Instance
+            for (i = 0; i < res[0].data_len; i++) {
+                if (res[0].data[i].id == server_id) {
+                    if (res[0].data[i].content.integer & rights_needed) {
+                        ret = true;
+                        goto exit;
+                    } else {
+                        ret = false;
+                        goto exit;
+                    }
+                }
+
+                //Keep the default ACL Resource Instance, if any, to save another loop later
+                if (res[0].data[i].id == DEFAULT_SHORT_SERVER_ID)
+                    default_acl = res[0].data[i].content.integer;
+            }
+
+            //If no ACL for this server, check if it is the owner of the object.
+            // If owner and no specific ACL Resource Instance, then full access rights.
+            if (res[1].data[0].content.integer == server_id) {
```

Comment:
Because spec [Pg. 60; Section 7.3.2.1 Obtaining Access Right].

I guess there could be a scenario in which I'm the owner, but my ACL exists and doesn't allow me to 'D'elete, for instance (maybe I changed it before, as the owner, or the bootstrap server changed it). If I check for owner first, I would be granted the access rights... but if I check for the ACL first (as the spec says), I would not. :)